runtime-sdk/src/subcall.rs

```diff
@@ -0,0 +1,214 @@
+//! Subcall dispatch.
+use std::{cell::RefCell, collections::BTreeMap};
+
+use crate::{
+    context::{BatchContext, Context, TxContext},
+    dispatcher,
+    module::CallResult,
+    modules::core::{Error, API as _},
+    runtime::Runtime,
+    storage::{current::TransactionResult, CurrentStore},
+    types::{token, transaction, transaction::CallerAddress},
+};
+
+thread_local! {
+    /// The subcall stack for tracking depth and other metadata.
+    static SUBCALL_STACK: RefCell<SubcallStack> = RefCell::new(SubcallStack::new());
+}
+
+/// Subcall validator.
+pub trait Validator {
+    /// Validate a subcall before it is performed.
+    fn validate(&self, info: &SubcallInfo) -> Result<(), Error>;
+}
+
+/// A validator which allows everything.
+pub struct AllowAllValidator;
```

Comment:
I guess the `subscall` dispatch will be used in other places in a future PR as well. Currently, it's only used in a single place, so the abstraction seems a bit arbitrary (e.g. the `Validator`  trait which is only implemented by the `AllowAllValidator` ).

Reply:
Yeah, see #1412 (WIP).